scene/resources/curve.cpp

```diff
@@ -1630,6 +1692,124 @@ Vector3 Curve3D::sample_baked(real_t p_offset, bool p_cubic) const {
 	}
 }
 
+Transform3D Curve3D::sample_backed_with_rotation(real_t p_offset, bool p_cubic, bool p_loop, Curve3D::RotationMode p_rotation_mode) const {
+	if (baked_cache_dirty) {
+		_bake();
+	}
+
+	// Validate: Curve may not have baked points.
+	const int point_count = baked_point_cache.size();
+	ERR_FAIL_COND_V_MSG(point_count == 0, Transform3D(), "No points in Curve3D.");
+
+	if (point_count == 1) {
+		Transform3D t;
+		t.origin = baked_point_cache.get(0);
+		ERR_FAIL_V_MSG(t, "Only 1 point in Curve3D.");
+	}
+
+	const real_t baked_length = get_baked_length();
+	if (baked_length == 0.0) {
+		Transform3D t;
+		t.origin = baked_point_cache.get(0);
+		ERR_FAIL_V_MSG(t, "Length is 0 in Curve3D.");
+	}
+
+	Vector3 pos = sample_baked(p_offset, p_cubic);
+
+	if (p_rotation_mode == ROTATION_NONE) {
+		Transform3D t;
+		t.origin = pos;
+		return t;
+	}
+
+	const real_t *dist_ptr = baked_dist_cache.ptr();
+	//const Vector3 *pos_ptr = baked_point_cache.ptr();
+	const real_t *tilt_ptr = baked_tilt_cache.ptr();
+
+	const Vector3 *up_ptr = baked_up_vector_cache.ptr();
+
+	const real_t offset_first = dist_ptr[0];
+	const real_t offset_last = dist_ptr[point_count - 1];
+	p_offset = CLAMP(p_offset, offset_first, offset_last); // or wrap?
+
+	int start = 0;
+	int end = point_count;
+	int idx = (end + start) / 2;
+	// Binary search to find baked points.
+	while (start < idx) {
+		real_t offset = baked_dist_cache[idx];
+		if (p_offset <= offset) {
+			end = idx;
+		} else {
+			start = idx;
+		}
+		idx = (end + start) / 2;
+	}
+
+	const real_t offset_begin = baked_dist_cache[idx];
+	const real_t offset_end = baked_dist_cache[idx + 1];
+
+	const real_t idx_interval = offset_end - offset_begin;
+	// redundant?
+	ERR_FAIL_COND_V_MSG(p_offset < offset_begin || p_offset > offset_end, Transform3D(), "failed to find baked segment");
+	const real_t frac = (p_offset - offset_begin) / idx_interval;
+
+	Vector3 forward_begin;
+	Vector3 forward_end;
+	if (idx == 0) {
+		forward_begin = (baked_point_cache[1] - baked_point_cache[0]).normalized();
+		forward_end = (baked_point_cache[1] - baked_point_cache[0]).normalized();
+	} else {
+		forward_begin = (baked_point_cache[idx] - baked_point_cache[idx - 1]).normalized();
+		forward_end = (baked_point_cache[idx + 1] - baked_point_cache[idx]).normalized();
+	}
+	const Vector3 up_begin = up_ptr[idx];
+	const Vector3 up_end = up_ptr[idx + 1];
+	const Vector3 sideway_begin = up_begin.cross(forward_begin);
+	const Vector3 sideway_end = up_end.cross(forward_end);
+
+	const Basis frame_begin = Basis(sideway_begin, up_begin, forward_begin); // Basis constructor is now colume major instead of row major
```

Comment:
```suggestion
	const Basis frame_begin = Basis(sideway_begin, up_begin, forward_begin); // Basis constructor is now column major instead of row major
```

Reply:
I refactor the code substantially, so I cannot commit it directly. But I corrected the function name during the refactor. Thanks for the review